query-service-impl/src/test/java/org/hypertrace/core/query/service/pinot/MigrationTest.java

```diff
@@ -233,36 +269,166 @@ public void testQueryWithDistinctCountAggregation() {
-        "SELECT tags__keys, tags__values FROM SpanEventView "
+        "SELECT mapValue(tags__keys,'span.kind',tags__values) FROM SpanEventView "
             + "WHERE "
             + viewDefinition.getTenantIdColumn()
             + " = '"
             + TENANT_ID
             + "' "
-            + "AND tags__keys = 'flags' and tags__values = '0' and mapvalue(tags__keys,'flags',tags__values) = '0'",
+            + "AND ( tags__keys = 'span.kind' and mapvalue(tags__keys,'span.kind',tags__values) > 'client' )",
         viewDefinition,
         executionContext);
   }
 
+  @Test
+  public void testQueryWithOrderByWithMapAttribute() {
+    Builder builder = QueryRequest.newBuilder();
+    Expression spanKind = createComplexAttributeExpression("Span.tags", "span.kind").build();
+    builder.addSelection(spanKind);
+
+    Filter greaterThanOrEqualToFilter =
+        Filter.newBuilder()
+            .setOperator(Operator.GE)
+            .setLhs(spanKind)
+            .setRhs(createStringLiteralValueExpression("client"))
+            .build();
+    builder.setFilter(
+        createCompositeFilter(
+            Operator.AND,
+            createContainsKeyFilter("Span.tags", "span.kind"),
+            greaterThanOrEqualToFilter));
+    builder.addOrderBy(createOrderByExpression(spanKind.toBuilder(), SortOrder.DESC));
+
+    ViewDefinition viewDefinition = getDefaultViewDefinition();
+    defaultMockingForExecutionContext();
+
+    assertPQLQuery(
+        builder.build(),
+        "select mapValue(tags__KEYS,'span.kind',tags__VALUES) FROM spanEventView "
+            + "WHERE "
+            + viewDefinition.getTenantIdColumn()
+            + " = '"
+            + TENANT_ID
+            + "' "
+            + "AND ( tags__keys = 'span.kind' and mapvalue(tags__keys,'span.kind',tags__values) >= 'client' ) "
+            + "order by mapvalue(tags__KEYS,'span.kind',tags__VALUES) "
+            + "DESC ",
+        viewDefinition,
+        executionContext);
+  }
+
+  @Test
+  public void testQueryWithGroupByWithMapAttribute() {
+    Builder builder = QueryRequest.newBuilder(buildGroupByMapAttributeQuery());
+    ViewDefinition viewDefinition = getDefaultViewDefinition();
+    defaultMockingForExecutionContext();
+    assertPQLQuery(
+        builder.build(),
+        "select mapValue(tags__KEYS,'span.kind',tags__VALUES), AVG(duration_millis) FROM spanEventView"
+            + " where "
+            + viewDefinition.getTenantIdColumn()
+            + " = '"
+            + TENANT_ID
+            + "' "
+            + "AND ( start_time_millis > 1570658506605 AND start_time_millis < 1570744906673 "
+            + "AND tags__keys = 'span.kind' "
+            + "AND mapValue(tags__KEYS,'span.kind',tags__VALUES) != '' ) "
+            + "group by mapValue(tags__KEYS,'span.kind',tags__VALUES)",
+        viewDefinition,
+        executionContext);
+  }
+
+  private QueryRequest buildGroupByMapAttributeQuery() {
+    Builder builder = QueryRequest.newBuilder();
```

Comment:
nit: For test inputs that are this complex, it's usually more readable and a better test (since it better mimics an upstream client) to provide them as text.

Reply:
I didn't get you

Reply:
Rather than building the query in code, this would mean deserializing it from text - either an inline string or a file.